syntax = "proto3";

package cadmus.pbuf.api.kleis.v2;

option go_package = "artemishealth.com/cadmus/proto/cadmus/pbuf/api/kleis/v2";

// The PermissionRequest protocol can be represented by a JSON of this form:
// {
//   "check_groups":[
//     {
//       "group_name":"group_permission_1",
//       "checks":[
//         {
//           "resource":{
//             "type":"TYPE_COMPANY",
//             "id":"rosebud"
//           },
//           "relation":"RELATION_WRITER",
//           "subject":{
//             "type":"TYPE_USER",
//             "id":"orsonwells"
//           }
//         }
//       ]
//     }
//   ]
// }
//
// A single permission check could alternatively be represented by this string:
// document:specificdocument#reader@user:specificuser
// |_______________________| |____| |_______________|
//         resource         relation     subject


message PermissionRequest {
  repeated CheckGroup check_groups = 1;
}

message CheckGroup {
  string group_name = 1;
  repeated Check checks = 2;
}

message Check {
  enum Relation {
    RELATION_UNSPECIFIED = 0;
    RELATION_READER = 1;
    RELATION_WRITER = 2;
    RELATION_USER = 3;
    RELATION_ADMIN = 4;
  }

  Resource resource = 1;
  Relation relation = 2;
  Subject subject = 3;
}


message Resource {
  enum Type {
    TYPE_UNSPECIFIED = 0;
    TYPE_COMPANY = 1;
    TYPE_APPLICATION = 2;
  }

  Type type = 1;
  string id = 2;
}

message Subject {
  enum Type {
    TYPE_UNSPECIFIED = 0;
    TYPE_COMPANY = 1;
    TYPE_USER = 2;
  }

  Type type = 1;
  string id = 2;
}